scapy/contrib/isotp.py

```diff
@@ -669,42 +741,56 @@ def __init__(self,
             listen_only=listen_only
         )
 
-        self.ins = impl
-        self.outs = impl
+        self.iso_ins = impl  # type: Optional[ISOTPSocketImplementation]
+        self.iso_outs = impl  # type: Optional[ISOTPSocketImplementation]
+        self.ins = cast(socket.socket, impl)
+        self.outs = cast(socket.socket, impl)
         self.impl = impl
-
+        self.basecls = basecls
         if basecls is None:
             warning('Provide a basecls ')
-        self.basecls = basecls
 
     def close(self):
+        # type: () -> None
         if not self.closed:
             self.impl.close()
+            self.iso_outs = None
+            self.iso_ins = None
             self.outs = None
-            self.ins = None
-            SuperSocket.close(self)
+            self.ins = None  # type: ignore
+            super(ISOTPSoftSocket, self).close()
 
     def begin_send(self, p):
+        # type: (Packet) -> int
         """Begin the transmission of message p. This method returns after
         sending the first frame. If multiple frames are necessary to send the
         message, this socket will unable to send other messages until either
         the transmission of this frame succeeds or it fails."""
         if hasattr(p, "sent_time"):
             p.sent_time = time.time()
 
-        return self.outs.begin_send(bytes(p))
+        if self.iso_outs is not None:
+            self.iso_outs.begin_send(bytes(p))
+            return len(p)
+        else:
+            return 0
 
     def recv_raw(self, x=0xffff):
+        # type: (int) -> Tuple[Optional[Type[Packet]], Optional[bytes], Optional[float]]  # noqa: E501
         """Receive a complete ISOTP message, blocking until a message is
         received or the specified timeout is reached.
         If self.timeout is 0, then this function doesn't block and returns the
         first frame in the receive buffer or None if there isn't any."""
-        msg = self.ins.recv()
-        t = time.time()
-        return self.basecls, msg, t
+        if self.iso_ins is not None:
+            return self.basecls, self.iso_ins.recv(), time.time()
```

Comment:
BTW is there nothing more precise than `time.time()` here? I'm not sure what kind of socket `iso_ins` is.

Reply:
Very good catch! I now use the time from the sockets which received the packet.